CIP-ledger-evolution/STS.md

```diff
@@ -0,0 +1,195 @@
+
+In describing a state transition system `L` we are principally concerned with six things:
+
+* **States** - The underlying datatype of our system.
+
+* **Transitions** - The means by which we might move from one (valid) state to another.
+
+* **Signals** - The means by which transitions are triggered.
+
+* **Predicates** - We do not prescribe a formal language with which to define predicates,
+but rather leave this up to the description of specific systems.
+Even so, a formal language may not been needed as long as the notation is clear and precise enough.
+Typically, finite set theory is all that is needed, with notation for concepts such as
+integers, sequences, powersets, functions, maps, comprehensions, conditionals, etc.
+Most of the predicates will have a value of true or false,
+but we will stretch the rules later when we define the composition of transition systems.
+
+* **Rules** - Judgments describing valid transitions. A rule has an antecedent
+(sometimes called premise) and a consequent (sometimes called conclusion), such that if
+the predicates in the antecedent hold, the consequent is assumed to be a valid state or transition.
+The predicates in the antecedent may contain variables for the state, the signal,
+and the environment (see below).
+
+* **Environment** - Sometimes we may implicitly rely on some additional information being present
+in the system to evaluate the rules. An environment does not have to exist (or, equivalently,
+we may have an empty environment), but crucially an environment is not modified by
+transitions.
+
+**Definition (State transition system)**
+
+A state transition system `L` is given by a 5-tuple `(S, T, Σ, R, Γ)`
+where:
+* `S` is a set of (not necessarily valid) states.
+* `Σ` is a set of signals.
+* `Γ` is a set of environment values.
+* `T` is a set of (not necessarily valid) transitions. We have that `T ⊆ 𝒫 (Γ × S × Σ × S)`.
+* `R` is a set of derivation rules.
+   A derivation rule is given by two parts:
+   * the antecedent, a predicate allowing values from `S ∪ Σ ∪ Γ`
+   * the consequent, a transition `t ∈ T`.
+* For every `s₀ ∈ S`, `σ ∈ S`, `γ ∈ Γ`, `r ∈ R`, if the antecedent of `r`
+  holds for `γ, s₀, σ`, then there is exactly one `s ∈ S` such that
+  `(γ, s₀, σ, s) ∈ T`.
+
+We refer to a 4-tuple `(γ, s₀, σ, s) ∈ T` as as **valid transition** if there is a rule `r ∈ R`
+such that the antecedent of `r` holds for `γ, s₀, σ`.
+
+## Notation
+
+We depict a derivation rule `r ∈ R` as follows:
+* There is a vertical line, with the antecedent above the line and the consequent below.
+* The consequent is depicted as follows, from left to right:
+  * The environment (if non-empty)
+  * The turnstile symbol, `⊢`
+  * The initial state (denoted `s₀` above)
+  * A right arrow with signal above it
+  * The new state (denoted `s` above)
+
+### Notational conveniences
+
+#### Multiple predicates
+
+Rather than a single logical predicate in the antecedent, we allow multiple predicates.
```

Comment:
Suggestion: allow standard boolean operators (`∧`,`∨`, ...)

Reply:
And say multiple predicates without connectors are implicitly connected by `and`

Reply:
not in a formal setting, probably